### PR TITLE
Upgrade an uploaded module when already installed

### PR DESCRIPTION
**Pull request by Quetzacoalt91**
| Questions | Answers |
| --- | --- |
| Branch? | develop |
| Description? | Currently, when we upload a module which is already installed, an error is thrown by the ModuleManager. In case of faillure, the ModuleController cleans the uploaded files, which are ... the module previously installed. This PR fixes the issue by launching an upgrade instead.<br/> At the same time, we introduce a ModuleZipManager to handle in only one place all the tasks about module name detection and saving. |
| Type? | bug fix |
| Category? | CO |
| BC breaks? | Nope |
| Deprecations? | Nope |
| Fixed ticket? | [BOOM-637](http://forge.prestashop.com/browse/BOOM-637) |
| How to test? | Upload an already installed module. You should not have the error 'this module is already installed' anymore. |

### CODE REVIEW

File: src/Adapter/Module/ModuleZipManager.php [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+     * @var Symfony\Component\Filesystem\Filesystem
+     */
+    private $filesystem;
+    /**
+     * @var Symfony\Component\Finder\Finder
+     */
+    private $finder;
+    private $translator;
+    
+    public function __construct(TranslatorInterface $translator)
```
Comment (mickaelandrieu) [inline src/Adapter/Module/ModuleZipManager.php]: can you put all dependencies as arguments of constructor ?

pros: easier to test.
cons: no cons ?

File: src/Adapter/Module/ModuleZipManager.php [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+        if (!Tools::ZipExtract($source, $sandboxPath)) {
+            throw new Exception(
+                $this->translator->trans(
+                    'Cannot extract module in %path% to get its name.',
+                    array('%path%' => $sandboxPath),
+                    'Admin.Modules.Notification'));
+        }
+
+        // Check the structure and get the module name
+        $directories = $this->finder->directories()
```
Comment (mickaelandrieu) [inline src/Adapter/Module/ModuleZipManager.php]: php-cs-fix :p

File: src/Adapter/Module/ModuleZipManager.php [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+                                    ->in($sandboxPath)
+                                    ->depth('== 0')
+                                    ->exclude(['__MACOSX'])
```
Comment (mickaelandrieu) [inline src/Adapter/Module/ModuleZipManager.php]: php-cs-fix :p

File: src/Adapter/Module/ModuleZipManager.php (hunk 1 above)
Comment (Quetzacoalt91) [inline src/Adapter/Module/ModuleZipManager.php]: One of them, finder, does not exist as a service.

Comment (mickaelandrieu) [inline src/Adapter/Module/ModuleZipManager.php]: you dont need to declare it as a service, only pass the FQCN :)

**Comment by mickaelandrieu** [comment]
Good job.

Dont forget to update your commit label ;)

Mickaël

**Comment by vTerenti** [comment]
Status : 'QA approved'

Best regards!

**Comment by mickaelandrieu** [comment]
Thank you @Quetzacoalt91

**Pull request merged**